Testing the Removal of Artificial Latency behind a feature flag

arafatkatze opened the pull request:
This PR aims to remove the artificial delay introduced for low-performance languages and specific AST nodes, as outlined in th [issue](https://linear.app/sourcegraph/issue/CODY-3315/delete-the-artificial-latency). The artificial delay was originally implemented for low performance languages so that users are not spammed with unhelpful suggestions. , but it has become less relevant over time, and it may now impact the predictability of autocomplete latency and user experience (UX).

# Changes:

## Introduce a feature flag:
A new feature flag has been added to toggle the artificial delay.
- Objective: The main objective is to achieve predictable latency and a consistent autocomplete UX for all languages.

- After merging this PR, the feature flag will be enabled for controlled testing, and data will be collected to assess the impact on CAR and latency for all languages
- Depending on the results, the team will decide whether to remove the artificial delay logic entirely.

## Test Plan:
I tested this PR locally by enabling and disabling the feature flag on my local Sourcegraph instance(I added the feature flag on it and then set the true/false values). I also used the debugger to run VS Code Extension on Desktop



## Changelog

<!-- OPTIONAL; info at https://www.notion.so/sourcegraph/Writing-a-changelog-entry-dd997f411d524caabf0d8d38a24a878c -->

dominiccooney commented on vscode/src/completions/artificial-delay.test.ts:
```diff
@@ -1,163 +1,191 @@
+
+        // css is a low performance language
+        const languageId = 'css'
+        expect(lowPerformanceLanguageIds.has(languageId)).toBe(true)
```
This unit testing style where you're asserting your assumptions is beautiful. Nice work.

dominiccooney commented on vscode/src/completions/artificial-delay.test.ts:
```diff
@@ -1,163 +1,191 @@
+            evaluateFeatureFlag: (flag: FeatureFlag) => Promise.resolve(enabledFeatureFlags.has(flag)),
+            refresh: () => {},
+            getFromCache: (flagName: FeatureFlag) => {
+                return false
```
Oooh this looks subtle or wrong. Is this used? If not, let's throw here to fail noisily. If so, why would we consult `enabledFeatureFlags` the first time, but not when we're checking the cache?

arafatkatze replied:
Yeah it was used elsewhere but not here so I removed it.

arafatkatze replied:
Full disclosure I was writing some other tests which entailed mocking the feature flags and those tests would hit [this part of the codebase](https://sourcegraph.com/-/editor?remote_url=https%3A%2F%2Fgithub.com%2Fsourcegraph%2Fcody.git&branch=artificial-latency-fix&file=vscode%2Fsrc%2Fcompletions%2Fcompletion-provider-config.ts&editor=VSCode&version=2.2.17&start_row=48&start_col=26&end_row=48&end_col=26)  where you see getFromCache method being called so I had to specifically mock out that thing(NOTE: It tries to get the feature flag value from cache and that behavior is slightly different from evaluateFeatureFlag which also calls [getFromCache](https://sourcegraph.com/-/editor?remote_url=https%3A%2F%2Fgithub.com%2Fsourcegraph%2Fcody.git&branch=artificial-latency-fix&file=lib%2Fshared%2Fsrc%2Fexperimentation%2FFeatureFlagProvider.ts&editor=VSCode&version=2.2.17&start_row=127&start_col=32&end_row=127&end_col=32) here). 

I kept that thing here because I thought it was necessary but turns out that its not so I got rid of that and started failing loudly.

dominiccooney reviewed (approved):
Is getting feature flags really fast, when they're cached? I'm just wondering if making `getFeatureFlag` async means you're going to introduce another kind of random delay that depends on how busy the main thread is.

Could we drill down on the point mentioned in the test?

valerybugakov commented on lib/shared/src/experimentation/FeatureFlagProvider.ts:
```diff
@@ -30,7 +30,8 @@ export enum FeatureFlag {
     CodyAutocompleteFIMModelExperimentVariant3 = 'cody-autocomplete-fim-model-experiment-variant-3-v2',
     CodyAutocompleteFIMModelExperimentVariant4 = 'cody-autocomplete-fim-model-experiment-variant-4-v2',
-
+    CodyInLineSuffixAutocomplete = 'cody-in-line-suffix-autocomplete',
```
This should live in a separate PR, right?

arafatkatze replied:
Sorry with a dozen rebases things get mixed up

valerybugakov commented on lib/shared/src/experimentation/FeatureFlagProvider.ts:
```diff
@@ -30,7 +30,8 @@ export enum FeatureFlag {
     CodyAutocompleteFIMModelExperimentVariant4 = 'cody-autocomplete-fim-model-experiment-variant-4-v2',
-
+    CodyInLineSuffixAutocomplete = 'cody-in-line-suffix-autocomplete',
+    CodyDisableArtificialDelay = 'cody-disable-artificial-delay',
```
Let's use the same `cody-autocomplete-x-y-z` pattern for feature flag names. Makes it easier to find all active autocomplete flags on DotCom in the site-admin.

arafatkatze replied:
Done

valerybugakov commented on vscode/src/completions/artificial-delay.ts:
```diff
@@ -42,12 +42,19 @@ let userMetrics = {
     completionIntent?: CompletionIntent
-): number {
+): Promise<number> {
+    const codyDisableArtificialDelay = await featureFlagProvider.instance!.evaluateFeatureFlag(
+        FeatureFlag.CodyDisableArtificialDelay
```
As Dominic pointed out, we should prefetch this feature using `vscode/src/completions/completion-provider-config.ts` and access the value here in a sync way.

valerybugakov commented on vscode/src/completions/completion-provider-config.ts:
```diff
@@ -161,6 +162,10 @@ class CompletionProviderConfig {
                 return 0
         }
     }
+
+    public get autocompleteEnableAritificialDelay(): boolean {
+        return this.getPrefetchedFlag(FeatureFlag.CodyAutocompleteDisableArtificialDelay)
+    }
```
Not used, can be deleted?

arafatkatze replied:
Fixed the logic in the new PR

valerybugakov commented on vscode/src/completions/artificial-delay.ts:
```diff
@@ -46,8 +45,12 @@ export function getArtificialDelay(
+    CodyAutocompleteDisableArtificialDelay: boolean,
     completionIntent?: CompletionIntent
 ): number {
+    if (CodyAutocompleteDisableArtificialDelay) {
```
See related logic below:
```ts
// Increase latency linearly up to max after 5 rejected suggestions
```

It's reasonable to keep using it. We just want to get rid of the baseline increase for low-perf languages and AST nodes.

arafatkatze replied:
Thanks for the clarification I made the. changes.

valerybugakov commented on vscode/src/completions/artificial-delay.ts:
```diff
@@ -46,8 +45,12 @@ export function getArtificialDelay(
     featureFlags: LatencyFeatureFlags,
     uri: string,
     languageId: string,
+    CodyAutocompleteDisableArtificialDelay: boolean,
```
Usually all the arguments use camelCase for naming.

arafatkatze replied:
Fixed

valerybugakov commented on vscode/src/completions/artificial-delay.test.ts:
```diff
@@ -196,21 +227,21 @@ describe('getArtificialDelay', () => {
         expect(lowPerformanceLanguageIds.has(goLanguageId)).toBe(false)
         // reset to provider latency because language latency is ignored for non-low-performance languages
-        expect(getArtificialDelay(featureFlagsNoUser, goUri, goLanguageId)).toBe(0)
+        expect(getArtificialDelay(featureFlagsNoUser, goUri, goLanguageId, false)).toBe(0)
```
Let's convert the argument list of an object to make it clear what `false` means.

arafatkatze replied:
Done

valerybugakov commented on vscode/src/completions/inline-completion-item-provider.ts:
```diff
@@ -414,11 +414,14 @@ export class InlineCompletionItemProvider
                 ),
             }
-
+            const CodyAutocompleteDisableArtificialDelay = completionProviderConfig.getPrefetchedFlag(
```
Nice!

valerybugakov commented on vscode/src/completions/artificial-delay.test.ts:
```diff
@@ -1,163 +1,194 @@
     })
 
-    it('returns gradually increasing latency up to max for CSS when suggestions are rejected', () => {
+    it('returns no artificial delay when CodyAutocompleteTracing feature flag is enabled', async () => {
```
`CodyAutocompleteTracing` — wrong flag name?

valerybugakov commented on vscode/src/completions/artificial-delay.test.ts:
```diff
@@ -1,163 +1,194 @@
+
     beforeEach(() => {
         vi.useFakeTimers()
+        originalFeatureFlagProviderInstance = featureFlagProvider.instance
+        enabledFeatureFlags.clear()
+        featureFlagProvider.instance = {
```
We can use `vi.spyOn` here to spy on relevant `featureFlagProvider` methods and replace their return values with `spy.mockReturnValue()`. Or we can mock the whole import by the path. With that in place, `vi.restoreAllMocks()` would make `featureFlagProvider.instance = originalFeatureFlagProviderInstance` obsolete.

Also, we can spy on the graphql client, which resolves feature flag values. That would test the feature flag provider logic, too.

arafatkatze replied:
@valerybugakov I got something better. I adopted the new logic to get the value of the feature flag [by doing a subscription](https://sourcegraph.com/-/editor?remote_url=https%3A%2F%2Fgithub.com%2Fsourcegraph%2Fcody.git&branch=artificial-latency-fix&file=vscode%2Fsrc%2Fcompletions%2Fcompletion-provider-config.ts&editor=VSCode&version=2.2.17&start_row=197&start_col=14&end_row=204&end_col=5)
 and then changed the function to have the args 
```

export function getArtificialDelay(params: {
    featureFlags: LatencyFeatureFlags
    uri: string
    languageId: string
    codyCompletionDisableLowPerfLangDelay: boolean
    completionIntent?: CompletionIntent
}): number {

```
This made the testing of the function much easier as I only have to tweak the params `codyCompletionDisableLowPerfLangDelay`. 


NOTE: 

1. The logic of subscription to a feature flag + GQL CLient is also tested [here](https://sourcegraph.com/-/editor?remote_url=https%3A%2F%2Fgithub.com%2Fsourcegraph%2Fcody.git&branch=artificial-latency-fix&file=lib%2Fshared%2Fsrc%2Fexperimentation%2FFeatureFlagProvider.test.ts&editor=VSCode&version=2.2.17&start_row=203&start_col=0&end_row=216&end_col=14)(but that's a separate thing).

valerybugakov replied:
This is great!

arafatkatze commented:
@valerybugakov Apologies I was still in the middle of fixing the PR when you saw the new changes. It wasn't in the right state(I will try to be mindful about communicating that the PR is not in the right state)

arafatkatze commented:
@valerybugakov I updated the PR again with all the recommendations that you suggested and now its a lot lot cleaner. 

With the latest changes from Quinn I have the subscription/update logic for feature flags also(Read[ this comment](https://github.com/sourcegraph/cody/pull/5480#discussion_r1760381822))

As for testing I have also incorporated the changes that you recommended so its easy to understand the tests too. 

Thanks again for such a thorough review, I would like another one now.

valerybugakov commented on vscode/src/completions/inline-completion-item-provider.ts:
```diff
@@ -435,13 +453,14 @@ export class InlineCompletionItemProvider
-                document.languageId,
-                completionIntent
-            )
+            const params = {
+                featureFlags: latencyFeatureFlags,
+                uri: document.uri.toString(),
+                languageId: document.languageId,
+                codyCompletionDisableLowPerfLangDelay: this.disableLowPerfLangDelay,
+                completionIntent: completionIntent,
+            }
+            const artificialDelay = getArtificialDelay({ ...params })
```
This way, the created object is type-checked on the declaration. Also, we don't need to create a new object from `params`.

```suggestion
            const artificialDelay = getArtificialDelay({
                featureFlags: latencyFeatureFlags,
                uri: document.uri.toString(),
                languageId: document.languageId,
                codyCompletionDisableLowPerfLangDelay: this.disableLowPerfLangDelay,
                completionIntent,
            })
```

valerybugakov commented on vscode/src/completions/inline-completion-item-provider.ts:
```diff
@@ -435,13 +453,14 @@ export class InlineCompletionItemProvider
-                completionIntent
-            )
+            const params = {
+                featureFlags: latencyFeatureFlags,
```
We can remove the `FeatureFlag.CodyAutocompleteUserLatency` feature flag and all the related logic. It's not actively used, and we do not plan to conduct new experiments with it. This can be done in a follow-up PR or here.

arafatkatze replied:
Will do it in a followup PR for sure.

valerybugakov commented on vscode/src/completions/inline-completion-item-provider.ts:
```diff
@@ -179,7 +181,23 @@ export class InlineCompletionItemProvider
         }
 
         this.requestManager = new RequestManager()
+        // Fetch the initial value for disableLowPerfLangDelay and set it
+        firstValueFrom(completionProviderConfig.completionDisableLowPerfLangDelay).then(
+            disableLowPerfLangDelay => {
+                this.disableLowPerfLangDelay = disableLowPerfLangDelay
+            }
+        )
 
+        // Subscribe to changes in disableLowPerfLangDelay and update the value accordingly
+        this.disposables.push(
+            subscriptionDisposable(
+                completionProviderConfig.completionDisableLowPerfLangDelay.subscribe(
+                    disableLowPerfLangDelay => {
+                        this.disableLowPerfLangDelay = disableLowPerfLangDelay
+                    }
+                )
+            )
+        )
```
Do we need both of these? There's an example of using `CodyAutocompleteTracing` in the `InlineCompletionItemProvider` constructor, which relies only on one call sequence to initialize and keep the local value up to date.

arafatkatze replied:
Both of these are technically NOT needed but I wanted to maintain the logical consistency with "initializing" something and then subscribing to its value but yeah since its redundant I don't see the reason to actually keep it for anything other than having a reasonably understandable logic.

valerybugakov commented on vscode/src/completions/artificial-delay.test.ts:
```diff
@@ -10,158 +9,218 @@ describe('getArtificialDelay', () => {
 
         const featureFlagsLangOnly = {
-            user: false,
+            user: codyCompletionDisableLowPerfLangDelay,
```
Is this a property for a different feature flag? `CodyAutocompleteUserLatency`

valerybugakov commented on vscode/src/completions/artificial-delay.test.ts:
```diff
@@ -171,46 +230,66 @@ describe('getArtificialDelay', () => {
     })
 
-    it('returns latency based on language only when user flag is disabled', () => {
+    it('returns latency based on language only when user flag is disabled', async () => {
```
Do we still need `async` callbacks here?

arafatkatze replied:
Good catch I removed them

valerybugakov commented on vscode/src/completions/artificial-delay.test.ts:
```diff
@@ -10,158 +9,218 @@ describe('getArtificialDelay', () => {
+        // expect no artificial delay when codyCompletionDisableLowPerfLangDelay is true
+        expect(getArtificialDelay(params)).toBe(0)
+    })
 
+    it('returns gradually increasing latency up to max for CSS when suggestions are rejected', async () => {
+        const uri = 'file://foo/bar/test.css'
+
+        // css is a low performance language
+        const languageId = 'css'
+        expect(lowPerformanceLanguageIds.has(languageId)).toBe(true)
+        const codyCompletionDisableLowPerfLangDelay = false
+        const params = {
+            featureFlags,
+            uri,
+            languageId,
+            codyCompletionDisableLowPerfLangDelay,
+        }
         // start with default high latency for low performance lang with default user latency added
-        expect(getArtificialDelay(featureFlags, uri, languageId)).toBe(1000)
-        expect(getArtificialDelay(featureFlags, uri, languageId)).toBe(1000)
-        expect(getArtificialDelay(featureFlags, uri, languageId)).toBe(1000)
-        expect(getArtificialDelay(featureFlags, uri, languageId)).toBe(1000)
-        expect(getArtificialDelay(featureFlags, uri, languageId)).toBe(1000)
+        expect(getArtificialDelay(params)).toBe(1000)
```
Nice!

valerybugakov commented on vscode/src/completions/artificial-delay.test.ts:
```diff
@@ -171,46 +228,66 @@ describe('getArtificialDelay', () => {
         const languageId = 'go'
         const goUri = 'foo/bar/test.go'
-        expect(lowPerformanceLanguageIds.has(languageId)).toBe(false)
+        expect(lowPerformanceLanguageIds.has(languageId)).toBe(codyAutocompleteDisableLowPerfLangDelay)
 
+        const params = {
+            featureFlags: featureFlagsLangOnly,
+            uri,
+            languageId,
+            codyAutocompleteDisableLowPerfLangDelay,
+        }
         // latency should only change based on language id when only the language flag is enabled
-        expect(getArtificialDelay(featureFlagsLangOnly, uri, lowPerformLanguageId)).toBe(1000)
-        expect(getArtificialDelay(featureFlagsLangOnly, uri, lowPerformLanguageId)).toBe(1000)
-        expect(getArtificialDelay(featureFlagsLangOnly, uri, lowPerformLanguageId)).toBe(1000)
-        expect(getArtificialDelay(featureFlagsLangOnly, uri, lowPerformLanguageId)).toBe(1000)
-        expect(getArtificialDelay(featureFlagsLangOnly, uri, lowPerformLanguageId)).toBe(1000)
-        expect(getArtificialDelay(featureFlagsLangOnly, uri, lowPerformLanguageId)).toBe(1000)
+        expect(getArtificialDelay({ ...params, languageId: lowPerformLanguageId })).toBe(1000)
+        expect(getArtificialDelay({ ...params, languageId: lowPerformLanguageId })).toBe(1000)
+        expect(getArtificialDelay({ ...params, languageId: lowPerformLanguageId })).toBe(1000)
+        expect(getArtificialDelay({ ...params, languageId: lowPerformLanguageId })).toBe(1000)
+        expect(getArtificialDelay({ ...params, languageId: lowPerformLanguageId })).toBe(1000)
+        expect(getArtificialDelay({ ...params, languageId: lowPerformLanguageId })).toBe(1000)
         // latency back to 0 when language is no longer low-performance
-        expect(getArtificialDelay(featureFlagsLangOnly, goUri, languageId)).toBe(0)
+        expect(getArtificialDelay({ ...params, uri: goUri, languageId })).toBe(0)
     })
 
     it('returns latency based on language only when user flag is disabled', () => {
         const uri = 'file://foo/bar/test.css'
 
         // css is a low performance language
         const languageId = 'css'
+        const codyAutocompleteDisableLowPerfLangDelay = false
         expect(lowPerformanceLanguageIds.has(languageId)).toBe(true)
 
         const featureFlagsNoUser = {
-            user: false,
+            user: codyAutocompleteDisableLowPerfLangDelay,
```
We need to update the variable name here because it doesn't match the underlying feature flag.

arafatkatze replied:
Done

valerybugakov reviewed (approved):
Thank you for addressing all the comments! 💜
